src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -197,41 +197,88 @@ public void GetItems<T>(ReadOnlySpan<T> choices, Span<T> destination)
-                while (!destination.IsEmpty)
+                // Get stack space to store random bytes. This size was chosen to balance between
+                // stack consumed and number of random calls required.
+                Span<byte> randomBytes = stackalloc byte[512];
+
+                if (BitOperations.IsPow2(choices.Length))
                 {
-                    if (destination.Length < randomBytes.Length)
+                    // To avoid bias, we can't just % all bytes to get them into range; that would cause
+                    // the lower values to be more likely than the higher values. If the number of choices
+                    // is a power of 2, though, we can just mask off the extraneous bits.
+
+                    int mask = choices.Length - 1;
+
+                    while (!destination.IsEmpty)
                     {
-                        randomBytes = randomBytes.Slice(0, destination.Length);
+                        // If this will be the last iteration, avoid over-requesting randomness.
+                        if (destination.Length < randomBytes.Length)
+                        {
+                            randomBytes = randomBytes.Slice(0, destination.Length);
+                        }
+
+                        NextBytes(randomBytes);
+
+                        for (int i = 0; i < randomBytes.Length; i++)
+                        {
+                            destination[i] = choices[randomBytes[i] & mask];
+                        }
+
+                        destination = destination.Slice(randomBytes.Length);
                     }
+                }
+                else
+                {
+                    // As the length isn't a power of two, we can't just mask off all extraneous bits, and
+                    // instead need to do rejection sampling. However, we can mask off the irrelevant bits, which
+                    // then reduces the chances of needing to reject a value.
 
-                    NextBytes(randomBytes);
+                    int mask = (int)BitOperations.RoundUpToPowerOf2((uint)choices.Length) - 1;
 
-                    int mask = choices.Length - 1;
-                    for (int i = 0; i < randomBytes.Length; i++)
+                    while (!destination.IsEmpty)
                     {
-                        destination[i] = choices[randomBytes[i] & mask];
+                        // Unlike in the IsPow2 case, where every byte will be used, some bytes here may
+                        // be rejected. On average, half the bytes may be rejected, so we heuristically
+                        // choose to shrink to twice the destination length.
+                        if (destination.Length * 2 < randomBytes.Length)
+                        {
+                            randomBytes = randomBytes.Slice(0, destination.Length * 2);
```

Comment:
Would it help if we increase the length a little to a multiple of 8 (`(destination.Length * 2 + 7) & ~7`)? It can increase the chance to get enough available bytes and doesn't hurt performance since `XoshiroImpl.NextBytes` generates 8 bytes a batch.

Reply:
That seems reasonable, thanks.